chanacceptor/rpcacceptor.go

```diff
@@ -0,0 +1,199 @@
+package chanacceptor
+
+import (
+	"fmt"
+	"sync"
+	"time"
+)
+
+const (
+	// DefaultRPCTimeout is the default time that the Accept method will
+	// listen for responses from the RPC server, after which it will timeout.
+	DefaultRPCTimeout = 15 * time.Second
+)
+
+// OpenChannelResponse is a struct containing the response from the RPC client
+// which includes whether or not they accept the open channel request and to
+// which PendingChannelID the response is to.
+type OpenChannelResponse struct {
+	// Accept is the decision from the RPC client that determines whether or not
+	// the client wishes to accept or reject the OpenChannel request.
+	Accept bool
+
+	// PendingChannelID represents the temporary id assigned to channels before
+	// they have a permanent channel id. This is used to determine which
+	// potential channel the above Accept applies to.
+	PendingChannelID [32]byte
+}
+
+// Client represents an RPC streaming client that responds to OpenChannel queries
+// issued by the RPCAcceptor.
+type Client struct {
+	// SendChan sends the *OpenChannelRequest to the RPC client.
+	SendChan chan *OpenChannelRequest
+
+	// RecvChan receives OpenChannelResponse from the RPC client.
+	RecvChan chan OpenChannelResponse
+
+	closeChan chan struct{}
+	active    bool
+}
+
+// RPCAcceptor represents the RPC-controlled variant of the ChannelAcceptor.
+// Only one RPC client is currently allowed at a time.
+type RPCAcceptor struct {
+	clientMtx sync.RWMutex
+	client    *Client
+
+	acceptChan       chan *OpenChannelRequest
+	acceptClientsMtx sync.RWMutex
+	acceptClients    map[[32]byte]chan bool
+
+	quit chan struct{}
+	wg   sync.WaitGroup
+}
+
+// Accept is a predicate on the OpenChannelRequest which is sent to the RPC
+// client who will respond with the ultimate decision. This assumes a registered
+// client exists.
+func (r *RPCAcceptor) Accept(req *OpenChannelRequest) bool {
+	// Register a new channel for the PendingChannelID associated with this
+	// OpenChannelRequest.
+	responseChan := make(chan bool)
+	r.acceptClientsMtx.Lock()
+	r.acceptClients[req.OpenChanMsg.PendingChannelID] = responseChan
+	r.acceptClientsMtx.Unlock()
+
+	select {
+	case r.acceptChan <- req:
+	case <-r.quit:
+		return false
+	}
+
+	select {
+	case accept, _ := <-responseChan:
+		return accept
+	case <-time.After(DefaultRPCTimeout):
+		return false
+	case <-r.quit:
+		return false
+	}
+}
+
+// RegisterClient registers an RPC streaming client. There may only be one RPC
+// streaming client at a time.
+func (r *RPCAcceptor) RegisterClient() (*Client, error) {
+
+	if r.active() {
+		return nil, fmt.Errorf("client already exists")
+	}
+
+	r.clientMtx.Lock()
+	r.client = &Client{
+		SendChan:  make(chan *OpenChannelRequest),
+		RecvChan:  make(chan OpenChannelResponse),
+		closeChan: make(chan struct{}),
+		active:    true,
+	}
+	r.clientMtx.Unlock()
+
+	return r.client, nil
+}
+
+// UnregisterClient unregisters the RPC streaming client.
```

Comment:
Will there ever be a case where there is more than one client?  How would that work?  If so, this method can then just accept a `ClientID` and call the `Unregister` method on the client whose `ClientID` matches the argument.

Reply:
I assumed more clients would just mean more chained acceptors. So if 4 clients are active, they must all accept for the channel to be accepted. This would allow more general operation, and also some interesting usecases around distributed acceptance.

Reply:
Yeah like threshold acceptance of channels in a sense.